### PR TITLE
Phx 712 policies2

### PR DESCRIPTION
**Pull request by psschroeter**
Policies. Services are mapped to resources. Methods are mapped to actions. Based upon the openapi.json. The "ResultType" isn't explicitly spelled out in there, so there's a little bit of guessing based upon the media type and swagger "definition" name to pull that back out instead of just using exactly what's in the swagger.

Preview of `rsc policy actions`
```bash
Resource          Action            Href
================= ================= =======================================================================
AppliedPolicy     index             /api/governance/projects/:project_id/applied_policies
                  create            /api/governance/projects/:project_id/applied_policies
                  show              /api/governance/projects/:project_id/applied_policies/:policy_id
                  delete            /api/governance/projects/:project_id/applied_policies/:policy_id
                  evaluate          /api/governance/projects/:project_id/applied_policies/:policy_id
                  show_log          /api/governance/projects/:project_id/applied_policies/:policy_id/log
                  show_status       /api/governance/projects/:project_id/applied_policies/:policy_id/status
----------------- ----------------- -----------------------------------------------------------------------
Incident          index             /api/governance/projects/:project_id/incidents
                  show              /api/governance/projects/:project_id/incidents/:incident_id
                  index_escalations /api/governance/projects/:project_id/incidents/:incident_id/escalations
                  index_resolutions /api/governance/projects/:project_id/incidents/:incident_id/resolutions
                  resolve           /api/governance/projects/:project_id/incidents/:incident_id
----------------- ----------------- -----------------------------------------------------------------------
PolicyTemplate    index             /api/governance/projects/:project_id/policy_templates
                  upload            /api/governance/projects/:project_id/policy_templates
                  compile           /api/governance/projects/:project_id/policy_templates
                  show              /api/governance/projects/:project_id/policy_templates/:template_id
                  update            /api/governance/projects/:project_id/policy_templates/:template_id
                  delete            /api/governance/projects/:project_id/policy_templates/:template_id
----------------- ----------------- -----------------------------------------------------------------------
PublishedTemplate index             /api/governance/orgs/:org_id/published_templates
                  create            /api/governance/orgs/:org_id/published_templates
                  show              /api/governance/orgs/:org_id/published_templates/:template_id
                  update            /api/governance/orgs/:org_id/published_templates/:template_id
                  delete            /api/governance/orgs/:org_id/published_templates/:template_id
                  hide              /api/governance/orgs/:org_id/published_templates/:template_id
                  unhide            /api/governance/orgs/:org_id/published_templates/:template_id
```

### CODE REVIEW

File: gen/goav2gen/definition.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+			oldFields = append(oldFields, f.Name)
+		}
+		for _, f := range dt.Fields {
+			newFields = append(newFields, f.Name)
+		}
+		use := "Old"
+		if len(newFields) > len(oldFields) {
+			use = "New"
+		}
+		if strings.Join(oldFields, ",") != strings.Join(newFields, ",") {
```
Comment (arangamani) [inline gen/goav2gen/definition.go]: What is this stuff? What is it comparing?

File: gen/goav2gen/helpers.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+func prettify(o interface{}) string {
+	s, err := json.MarshalIndent(o, "", "    ")
+	if err != nil {
+		return fmt.Sprintf("%+v", o)
+	}
+	return string(s)
+}
+
+// cleans up description, removing blank lines and extraneous info
+func cleanDescription(doc string) string {
```
Comment (arangamani) [inline gen/goav2gen/helpers.go]: It might be useful to have the required security scopes in showing help.

File: policy/codegen_client.go [hunk 3]
```diff
@@ -0,0 +1,1351 @@
...
+}
+
+// AppliedPolicyLocator builds a locator from the given href.
+func (api *API) AppliedPolicyLocator(href string) *AppliedPolicyLocator {
+	return &AppliedPolicyLocator{Href(href), api}
+}
+
+//===== Actions
+
+// get /api/governance/projects/{project_id}/applied_policies
```
Comment (arangamani) [inline policy/codegen_client.go]: Can we capitalize the verb/method?

File: gen/goav2gen/definition.go (hunk 1 above)
Comment (psschroeter) [reply gen/goav2gen/definition.go]: Ah so this is the downside of swagger 2 vs swagger 3. For swagger 2 if you have multiple things a given type like an applied policy or user they all have to create their own definition for it in the swagger "definitions" section. So they'll be X copies of identical definition for a User struct lets say. This tries to collapse those things down into one return type struct. Apparently swagger 3 reorgs things a bit so they'll all point to the same thing if they are the same thing as there won't be any guess work.

File: gen/goav2gen/helpers.go (hunk 2 above)
Comment (psschroeter) [reply gen/goav2gen/helpers.go]: Yeah could go either way -- I think I'd rather link to a page describing the roles/scopes as a whole maybe as a blurb at the top, the rsc --help command is a bit too cluttered right now with every product thrown in at the same time.

File: policy/codegen_client.go (hunk 3 above)
Comment (psschroeter) [reply policy/codegen_client.go]: K

**Comment by psschroeter** [comment]
Yeah was going to get around to adding etag as header but didn't quite do it as I'd have to delve into various other parts and change them as well.

**Pull request merged**